tracer/src/Datadog.Trace/AppSec/Waf/ReturnTypes.Managed/InitializationResult.cs

```diff
@@ -0,0 +1,91 @@
+// <copyright file="InitializationResult.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using Datadog.Trace.AppSec.Waf.NativeBindings;
+using Datadog.Trace.Vendors.Newtonsoft.Json;
+
+namespace Datadog.Trace.AppSec.Waf.ReturnTypesManaged
+{
+    internal class InitializationResult
+    {
+        public InitializationResult(IntPtr? ruleHandle, ushort failedToLoadRules, ushort loadedRules, string ruleFileVersion, IReadOnlyDictionary<string, string[]> errors, bool unusableRuleFile = false)
+        {
+            HasErrors = errors.Count > 0;
+            Errors = errors;
+            ErrorMessage = HasErrors ? JsonConvert.SerializeObject(errors) : string.Empty;
+            FailedToLoadRules = failedToLoadRules;
+            LoadedRules = loadedRules;
+            RuleFileVersion = ruleFileVersion;
+            RuleHandle = ruleHandle;
+            UnusableRuleFile = unusableRuleFile;
+        }
+
+        internal IntPtr? RuleHandle { get; }
+
+        internal ushort FailedToLoadRules { get; }
+
+        /// <summary>
+        /// Gets the number of rules successfully loaded
+        /// </summary>
+        internal ushort LoadedRules { get; }
+
+        internal IReadOnlyDictionary<string, string[]> Errors { get; }
+
+        internal string ErrorMessage { get; }
+
+        internal bool HasErrors { get; }
+
+        internal bool UnusableRuleFile { get; }
+
+        internal string RuleFileVersion { get; }
+
+        internal string File { get; set; }
+
+        internal bool FileNotFound { get; set; }
+
+        internal bool ExceptionDecodingFile { get; set; }
+
+        internal static InitializationResult FromUnusableRuleFile() => new(null, 0, 0, string.Empty, new Dictionary<string, string[]>(), true);
+
+        internal static InitializationResult From(DdwafRuleSetInfoStruct ddwaRuleSetInfoStruct, IntPtr? ruleHandle)
+        {
+            var ddwafObjectStruct = ddwaRuleSetInfoStruct.Errors;
+            var errors = Decode(ddwafObjectStruct);
+            var ruleFileVersion = Marshal.PtrToStringAnsi(ddwaRuleSetInfoStruct.Version);
+            return new(ruleHandle, ddwaRuleSetInfoStruct.Failed, ddwaRuleSetInfoStruct.Loaded, ruleFileVersion, errors);
+        }
+
+        private static IReadOnlyDictionary<string, string[]> Decode(DdwafObjectStruct ddwafObjectStruct)
+        {
+            var errorsDic = new Dictionary<string, string[]>();
```

Comment:
It looks like you may be able to initialize this to with the required capacity.
```suggestion
            var errorsDic = new Dictionary<string, string[]>((int)ddwafObjectStruct.NbEntries);
```

Reply:
added in [420896c](https://github.com/DataDog/dd-trace-dotnet/pull/2638/commits/420896c7587436e018d62ae2534a27935cda878f)